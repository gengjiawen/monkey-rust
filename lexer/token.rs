use serde::{Deserialize, Serialize};
use std::fmt;
use std::fmt::Formatter;

#[derive(Clone, Debug, Eq, Hash, Ord, Serialize, Deserialize, PartialOrd, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(Clone, Debug, Eq, Hash, Ord, Serialize, Deserialize, PartialOrd, PartialEq)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        write!(f, "start: {}, end: {}, kind: {}", self.span.start, self.span.end, self.kind)
    }
}

#[derive(Clone, Debug, Eq, Hash, Ord, Serialize, Deserialize, PartialOrd, PartialEq)]
#[serde(tag = "type", content = "value")]
pub enum TokenKind {
    ILLEGAL,
    EOF,

    // Identifiers + literals
    IDENTIFIER { name: String },
    INT(i64),
    STRING(String),

    // Operators
    ASSIGN,   // =
    PLUS,     // +
    MINUS,    // -
    BANG,     // !
    ASTERISK, // *
    SLASH,    // /

    LT, // <
    GT, // >

    EQ,    // ==
    NotEq, // !=

    // delimiters
    COMMA,
    SEMICOLON,
    COLON,

    LPAREN,
    RPAREN,
    LBRACE,   // {
    RBRACE,   // }
    LBRACKET, // [
    RBRACKET, // ]

    // keywords
    FUNCTION,
    LET,
    TRUE,
    FALSE,
    IF,
    ELSE,
    RETURN,
}

pub fn lookup_identifier(identifier: &str) -> TokenKind {
    match identifier {
        "fn" => TokenKind::FUNCTION,
        "let" => TokenKind::LET,
        "true" => TokenKind::TRUE,
        "false" => TokenKind::FALSE,
        "if" => TokenKind::IF,
        "else" => TokenKind::ELSE,
        "return" => TokenKind::RETURN,
        _ => TokenKind::IDENTIFIER { name: identifier.to_string() },
    }
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            TokenKind::IDENTIFIER { name } => write!(f, "{}", name),
            TokenKind::INT(i) => write!(f, "{}", i),
            TokenKind::STRING(s) => write!(f, "{}", s),
            TokenKind::ASSIGN => write!(f, "="),
            TokenKind::PLUS => write!(f, "+"),
            TokenKind::MINUS => write!(f, "-"),
            TokenKind::BANG => write!(f, "!"),
            TokenKind::ASTERISK => write!(f, "*"),
            TokenKind::SLASH => write!(f, "/"),
            TokenKind::LT => write!(f, "<"),
            TokenKind::GT => write!(f, ">"),
            TokenKind::EQ => write!(f, "=="),
            TokenKind::NotEq => write!(f, "!="),
            TokenKind::COMMA => write!(f, ","),
            TokenKind::SEMICOLON => write!(f, ";"),
            TokenKind::LPAREN => write!(f, "("),
            TokenKind::RPAREN => write!(f, ")"),
            TokenKind::LBRACE => write!(f, "{{"),
            TokenKind::RBRACE => write!(f, "}}"),
            TokenKind::LBRACKET => write!(f, "["),
            TokenKind::RBRACKET => write!(f, "]"),
            TokenKind::FUNCTION => write!(f, "fn"),
            TokenKind::LET => write!(f, "let"),
            TokenKind::TRUE => write!(f, "true"),
            TokenKind::FALSE => write!(f, "false"),
            TokenKind::IF => write!(f, "if"),
            TokenKind::ELSE => write!(f, "else"),
            TokenKind::RETURN => write!(f, "return"),
            TokenKind::ILLEGAL => write!(f, "ILLEGAL"),
            TokenKind::EOF => write!(f, "EOF"),
            TokenKind::COLON => write!(f, ":"),
        }
    }
}
