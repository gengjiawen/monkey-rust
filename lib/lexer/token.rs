
#[derive(Clone, Debug, Eq, Hash, Ord, PartialOrd, PartialEq)]
pub enum Token {
    ILLEGAL,
    EOF,

    // Identifiers + literals
    IDENT(String),
    Int(i64),

    // Operators
    ASSIGN,
    PLUS,

    // delimiters
    COMMA,
    SEMICOLON,

    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,

    // keywords
    FUNCTION,
    LET
}

