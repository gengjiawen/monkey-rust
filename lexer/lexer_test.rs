#[cfg(test)]
mod tests {
    use crate::token::{Token, TokenKind};
    use crate::Lexer;
    use insta::*;

    fn test_token_set(l: &mut Lexer) -> Vec<Token> {
        let mut token_vs: Vec<Token> = vec![];
        loop {
            let t = l.next_token();
            if t.kind == TokenKind::EOF {
                token_vs.push(t);
                break;
            } else {
                token_vs.push(t);
            }
        }
        token_vs
    }

    pub fn test_lexer_common(name: &str, input: &str) {
        let mut l = Lexer::new(input);
        let token_vs = test_token_set(&mut l);

        assert_snapshot!(
            name,
            serde_json::to_string_pretty(&token_vs).unwrap(),
            input
        );
    }

    #[test]
    fn test_lexer_simple() {
        test_lexer_common("simple", "=+(){},:;");
    }

    #[test]
    fn test_lexer_let() {
        test_lexer_common("let", "let x=5");
    }

    #[test]
    fn test_comments() {
        test_lexer_common("comments", "// I am comments");
    }

    #[test]
    fn test_lexer_let_with_space() {
        test_lexer_common("let_with_space", "let x = 5");
    }

    #[test]
    fn test_lexer_string() {
        test_lexer_common("string", r#""a""#);
    }

    #[test]
    fn test_lexer_array() {
        test_lexer_common("array", "[3]");
    }

    #[test]
    fn test_lexer_hash() {
        test_lexer_common("hash", r#"{"one": 1, "two": 2, "three": 3}"#);
    }

    #[test]
    fn test_lexer_bool() {
        test_lexer_common("bool", "let y=true");
    }

    #[test]
    fn test_lexer_complex() {
        test_lexer_common(
            "complex",
            "
// welcome to monkeylang
let five = 5;
let ten = 10;

let add = fn(x, y) {
  x + y;
};

let result = add(five, ten);
!-/*5;
5 < 10 > 5;

if (5 < 10) {
	return true;
} else {
	return false;
}

10 == 10;
10 != 9;",
        );
    }
}
