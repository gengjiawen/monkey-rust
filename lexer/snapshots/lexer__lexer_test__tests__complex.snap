---
source: lexer/lexer_test.rs
expression: "\n// welcome to monkeylang\nlet five = 5;\nlet ten = 10;\n\nlet add = fn(x, y) {\n  x + y;\n};\n\nlet result = add(five, ten);\n!-/*5;\n5 < 10 > 5;\n\nif (5 < 10) {\n\treturn true;\n} else {\n\treturn false;\n}\n\n10 == 10;\n10 != 9;"
---
[
  {
    "kind": {
      "type": "LET"
    },
    "span": {
      "start": 26,
      "end": 29
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "five"
      }
    },
    "span": {
      "start": 30,
      "end": 34
    }
  },
  {
    "kind": {
      "type": "ASSIGN"
    },
    "span": {
      "start": 35,
      "end": 36
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 5
    },
    "span": {
      "start": 37,
      "end": 38
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 38,
      "end": 39
    }
  },
  {
    "kind": {
      "type": "LET"
    },
    "span": {
      "start": 40,
      "end": 43
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "ten"
      }
    },
    "span": {
      "start": 44,
      "end": 47
    }
  },
  {
    "kind": {
      "type": "ASSIGN"
    },
    "span": {
      "start": 48,
      "end": 49
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 10
    },
    "span": {
      "start": 50,
      "end": 52
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 52,
      "end": 53
    }
  },
  {
    "kind": {
      "type": "LET"
    },
    "span": {
      "start": 55,
      "end": 58
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "add"
      }
    },
    "span": {
      "start": 59,
      "end": 62
    }
  },
  {
    "kind": {
      "type": "ASSIGN"
    },
    "span": {
      "start": 63,
      "end": 64
    }
  },
  {
    "kind": {
      "type": "FUNCTION"
    },
    "span": {
      "start": 65,
      "end": 67
    }
  },
  {
    "kind": {
      "type": "LPAREN"
    },
    "span": {
      "start": 67,
      "end": 68
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "x"
      }
    },
    "span": {
      "start": 68,
      "end": 69
    }
  },
  {
    "kind": {
      "type": "COMMA"
    },
    "span": {
      "start": 69,
      "end": 70
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "y"
      }
    },
    "span": {
      "start": 71,
      "end": 72
    }
  },
  {
    "kind": {
      "type": "RPAREN"
    },
    "span": {
      "start": 72,
      "end": 73
    }
  },
  {
    "kind": {
      "type": "LBRACE"
    },
    "span": {
      "start": 74,
      "end": 75
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "x"
      }
    },
    "span": {
      "start": 78,
      "end": 79
    }
  },
  {
    "kind": {
      "type": "PLUS"
    },
    "span": {
      "start": 80,
      "end": 81
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "y"
      }
    },
    "span": {
      "start": 82,
      "end": 83
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 83,
      "end": 84
    }
  },
  {
    "kind": {
      "type": "RBRACE"
    },
    "span": {
      "start": 85,
      "end": 86
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 86,
      "end": 87
    }
  },
  {
    "kind": {
      "type": "LET"
    },
    "span": {
      "start": 89,
      "end": 92
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "result"
      }
    },
    "span": {
      "start": 93,
      "end": 99
    }
  },
  {
    "kind": {
      "type": "ASSIGN"
    },
    "span": {
      "start": 100,
      "end": 101
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "add"
      }
    },
    "span": {
      "start": 102,
      "end": 105
    }
  },
  {
    "kind": {
      "type": "LPAREN"
    },
    "span": {
      "start": 105,
      "end": 106
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "five"
      }
    },
    "span": {
      "start": 106,
      "end": 110
    }
  },
  {
    "kind": {
      "type": "COMMA"
    },
    "span": {
      "start": 110,
      "end": 111
    }
  },
  {
    "kind": {
      "type": "IDENTIFIER",
      "value": {
        "name": "ten"
      }
    },
    "span": {
      "start": 112,
      "end": 115
    }
  },
  {
    "kind": {
      "type": "RPAREN"
    },
    "span": {
      "start": 115,
      "end": 116
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 116,
      "end": 117
    }
  },
  {
    "kind": {
      "type": "BANG"
    },
    "span": {
      "start": 118,
      "end": 119
    }
  },
  {
    "kind": {
      "type": "MINUS"
    },
    "span": {
      "start": 119,
      "end": 120
    }
  },
  {
    "kind": {
      "type": "SLASH"
    },
    "span": {
      "start": 120,
      "end": 121
    }
  },
  {
    "kind": {
      "type": "ASTERISK"
    },
    "span": {
      "start": 121,
      "end": 122
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 5
    },
    "span": {
      "start": 122,
      "end": 123
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 123,
      "end": 124
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 5
    },
    "span": {
      "start": 125,
      "end": 126
    }
  },
  {
    "kind": {
      "type": "LT"
    },
    "span": {
      "start": 127,
      "end": 128
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 10
    },
    "span": {
      "start": 129,
      "end": 131
    }
  },
  {
    "kind": {
      "type": "GT"
    },
    "span": {
      "start": 132,
      "end": 133
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 5
    },
    "span": {
      "start": 134,
      "end": 135
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 135,
      "end": 136
    }
  },
  {
    "kind": {
      "type": "IF"
    },
    "span": {
      "start": 138,
      "end": 140
    }
  },
  {
    "kind": {
      "type": "LPAREN"
    },
    "span": {
      "start": 141,
      "end": 142
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 5
    },
    "span": {
      "start": 142,
      "end": 143
    }
  },
  {
    "kind": {
      "type": "LT"
    },
    "span": {
      "start": 144,
      "end": 145
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 10
    },
    "span": {
      "start": 146,
      "end": 148
    }
  },
  {
    "kind": {
      "type": "RPAREN"
    },
    "span": {
      "start": 148,
      "end": 149
    }
  },
  {
    "kind": {
      "type": "LBRACE"
    },
    "span": {
      "start": 150,
      "end": 151
    }
  },
  {
    "kind": {
      "type": "RETURN"
    },
    "span": {
      "start": 153,
      "end": 159
    }
  },
  {
    "kind": {
      "type": "TRUE"
    },
    "span": {
      "start": 160,
      "end": 164
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 164,
      "end": 165
    }
  },
  {
    "kind": {
      "type": "RBRACE"
    },
    "span": {
      "start": 166,
      "end": 167
    }
  },
  {
    "kind": {
      "type": "ELSE"
    },
    "span": {
      "start": 168,
      "end": 172
    }
  },
  {
    "kind": {
      "type": "LBRACE"
    },
    "span": {
      "start": 173,
      "end": 174
    }
  },
  {
    "kind": {
      "type": "RETURN"
    },
    "span": {
      "start": 176,
      "end": 182
    }
  },
  {
    "kind": {
      "type": "FALSE"
    },
    "span": {
      "start": 183,
      "end": 188
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 188,
      "end": 189
    }
  },
  {
    "kind": {
      "type": "RBRACE"
    },
    "span": {
      "start": 190,
      "end": 191
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 10
    },
    "span": {
      "start": 193,
      "end": 195
    }
  },
  {
    "kind": {
      "type": "EQ"
    },
    "span": {
      "start": 197,
      "end": 198
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 10
    },
    "span": {
      "start": 199,
      "end": 201
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 201,
      "end": 202
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 10
    },
    "span": {
      "start": 203,
      "end": 205
    }
  },
  {
    "kind": {
      "type": "NotEq"
    },
    "span": {
      "start": 207,
      "end": 208
    }
  },
  {
    "kind": {
      "type": "INT",
      "value": 9
    },
    "span": {
      "start": 209,
      "end": 210
    }
  },
  {
    "kind": {
      "type": "SEMICOLON"
    },
    "span": {
      "start": 210,
      "end": 211
    }
  },
  {
    "kind": {
      "type": "EOF"
    },
    "span": {
      "start": 211,
      "end": 212
    }
  }
]
